*: use a human friendly ID to tag and name cloud resources.

abhinavdahiya opened the pull request:
AWS:
- All the resources are named `<cluster_id>-<something>` and all names use `-` for separator.
This helps with consistently naming all the resources, and keeping the cluster_id in the reduces the probability of collisions.

- All tags include `Name` key with name mathcing the name for the resource.
This allows cluster to find resources using tags based on names.
Also move some of the `Name` tags from aws root module into specific modules because setting `Name` at root level makes it difficult for overriding the `Name` locally.
- All resources are tagged with `kubernetes.io/cluster/<cluster_id>: owned`

The loadbalancer target group names also have a max length of 32, therefore the target groups are using 4 character long suffixes.

Libvirt:
- All resources are prefixed with `cluster_id`

Openstack:
- all resources are prefixed with `cluster_id`
- openstack continues to use `openshiftClusterID: cluster_id` to tag all its resources.

The machine-api requires that all machine objects reference back to cluster object as specified by `sigs.k8s.io/cluster-api-cluster` label
on the machine objects to make sure masters are correctly adopted. Therfore the name for cluster object is the id used to tag the resources.

staebler commented on data/data/aws/bootstrap/variables.tf:
```diff
@@ -3,9 +3,9 @@ variable "ami" {
+variable "cluster_id" {
   type        = "string"
-  description = "The name of the cluster."
+  description = "The identifier for the cluster."
```
Can we add human-readable to the description to increase the differentiation between the UUID?

abhinavdahiya replied:
Why should we differentiate between UUID and HumanID in terraform?

staebler replied:
Within terraform we do not need to differentiate. However, this is at the boundary between terraform and go. The go code is ambiguous with the term cluster ID. As a reader of the code, it is not immediately clear to me here whether the cluster ID that terraform is expecting is the UUID or the human ID. We could either (1) expect the reader of the code to dig further into the code to figure out which one is expected here or (2) add more information to the description of the variable to make it clear.

abhinavdahiya replied:
the boundary says give me an identifier for the cluster, it doesn't need to care if it was given an UUID or a ID that is human readable... ?

Let me update it to include constraints like length, characters,  because that's what it cares any nothing more.

abhinavdahiya replied:
https://github.com/openshift/installer/pull/1280/commits/e0614f224730024ea22c3e4edee053f617970739

wking commented on pkg/asset/cluster/tfvars.go:
```diff
@@ -79,7 +79,7 @@ func (t *TerraformVariables) Generate(parents asset.Parents) error {
 	data, err := tfvars.TFVars(
-		clusterID.ClusterID,
+		clusterID.HumanID,
 		installConfig.Config.ObjectMeta.Name,
```
Can we drop this?  A non-uniquified name seems like something we'd never use.

abhinavdahiya replied:
https://github.com/openshift/installer/pull/1280/commits/3711b75d6bf53e2c7bfd0e15af7d5044b06cebbf

wking commented on data/data/aws/bootstrap/main.tf:
```diff
@@ -36,7 +40,7 @@ resource "aws_iam_instance_profile" "bootstrap" {
 
 resource "aws_iam_role" "bootstrap" {
-  name = "${var.cluster_name}-bootstrap-role"
+  name = "${var.cluster_id}-bootstrap-role"
```
While we're cleaning up, I think we can drop `-role`, which is already implied by the resource type.  Basically, every bootstrap resource can use `${var.cluster_id}-bootstrap` unless we create several of a single resource type.

abhinavdahiya replied:
I think it is fine if we keep the names as is. it doesn't hurt here as `iam roles` do not seem to have any limitation on length of name.

staebler commented on pkg/tfvars/tfvars.go:
```diff
@@ -8,7 +8,7 @@ import (
 type config struct {
 	ClusterID   string `json:"cluster_id,omitempty"`
-	Name        string `json:"cluster_name,omitempty"`
+	ClusterName string `json:"cluster_name,omitempty"`
```
I think it would be worthwhile to replace this with `ClusterDomain`. That way the logic for how the cluster domain is constructed is contained in only one place rather than in both go and terraform.

abhinavdahiya replied:
done with https://github.com/openshift/installer/pull/1280/commits/3711b75d6bf53e2c7bfd0e15af7d5044b06cebbf

wking commented on data/data/aws/main.tf:
```diff
@@ -16,29 +16,24 @@ module "bootstrap" {
-  tags = "${merge(map(
-      "Name", "${var.cluster_name}-bootstrap",
-    ), local.tags)}"
+  tags = "${local.tags}"
```
See 8a37f72 and f56c341 about not putting `kubernetes.io/cluster/...` on bootstrap resources except the instance.  Do you feel like the accidental-adoption concern is ungrounded?

abhinavdahiya replied:
everybidy needs to have ``kubernetes.io/cluster/...`` tag so that destroy captures bootstrap resources.

staebler replied:
Are we still planning to use the `kubernetes.io/cluster/...` tag for destroy? Since everything is tagged with the `Name` tag, can't we search on that instead?

abhinavdahiya replied:
> Are we still planning to use the `kubernetes.io/cluster/...` tag for destroy? Since everything is tagged with the `Name` tag, can't we search on that instead?

`Name` is resource name and has nothing to do with the cluster. AWS will use `kubernetes.io/cluster/...` tag.

wking replied:
I'm in favor of keeping a tag that Kubernetes ignores (like `openshiftClusterID`) for tearing down bootstrap resources that are not part of the Kubernetes cluster.  But if folks think we don't have to worry about the cluster accidentally adopting bootstrap resources, I'm ok collapsing to just the `kubernetes.io/cluster/...` tag.

wking commented on data/data/aws/vpc/master-elb.tf:
```diff
@@ -1,38 +1,44 @@
 
-  tags = "${var.tags}"
+  tags = "${merge(map(
+    "Name", "${var.cluster_id}-api-int",
```
Why set `Name` on resources that have a `name` property, if we're setting the same value in both?  I think it only makes sense when we feel like `name` must be overly truncated and want the tag for the unabbreviated version.

abhinavdahiya replied:
from https://github.com/openshift/installer/pull/1280/commits/d6de59aa39811ee0511998a8d251837026ac8ef5
```
- All tags include `Name` key with name mathcing the name for the resource.
This allows cluster to find resources using tags based on names.
```

wking commented on pkg/asset/installconfig/clusterid.go:
```diff
@@ -1,26 +1,50 @@
+	dep.Get(ica)
+	base := ica.Config.ObjectMeta.Name
+	if len(base) > maxBaseLen {
+		base = base[:maxBaseLen]
```
I think we want to trim sensitive characters, see #1089.

wking replied:
Hmm, this is also imposing AWS-specific length and (eventually) charset restrictions on a nominally-platform-agnostic property.  Will this end up as a race to the bottom?  Maybe it should be a platform-specific asset?  Or do we expect this to be something almost all platforms will need?

staebler replied:
I considered this same thing with the PR that I put out earlier. My decision was that for the time being there is no reason to make the HumanID characteristics unique to the platform as we do not have any platforms currently that place more onerous restrictions than we face with ELBs from AWS. It should be easy enough to change to platform-specific HumanIDs in the future if and when they are beneficial.

abhinavdahiya replied:
Yeah, we should be able to send cluste_ids to different platforms if that becomes a problem.
IMO this HumanID has enough constraints to work through most.

> I think we want to trim sensitive characters too, see #1089.

Will do.

abhinavdahiya replied:
> > I think we want to trim sensitive characters too, see #1089.
> 
> Will do.

added here https://github.com/openshift/installer/pull/1280/commits/55b241004756c2b9636f6527196ed3bc77a6d094

staebler commented on data/data/aws/main.tf:
```diff
@@ -106,6 +94,10 @@ resource "aws_route53_record" "etcd_cluster" {
   records = ["${formatlist("0 10 2380 %s", aws_route53_record.etcd_a_nodes.*.fqdn)}"]
 }
 
+data "aws_route53_zone" "public" {
```
What is the motivation behind bringing this here from the route53 module? I would have expected that we would want to go the other way with these route53 resources and bring `aws_route53_record.etcd_a_nodes`, `aws_route53_record.etcd_cluster`, and `aws_route53_zone.int` to the route53 module instead of bringing route53 resources out of the route53 module.

abhinavdahiya replied:
with how the current `r53` module is setup it takes private zone id and base_domain, and I think pulling the public zone out makes it so that the module takes ids for both for better consistency.

from https://github.com/openshift/installer/pull/1280/commits/d6de59aa39811ee0511998a8d251837026ac8ef5
```
Also moves the public zone discover to root module and pass the `public_route53_id`, similar to private zone, to `route53` zone.
```

and for moving the etcd records into the module, that makes sense. let me see and if the change is simple i'll do that.

staebler replied:
Well, my contention is that `aws_route32_zone.int` would be better placed in the route53 module. If that were moved, then there would be consistency in that neither public zone id nor private zone id would be passed into the route53 module.

abhinavdahiya replied:
How would you then turn off either of them (not that you can do them right now..)

staebler replied:
I don't know. How would you turn off either of them if they were in main instead of the route53 module? How would that be different if they were in the route53 module?

abhinavdahiya replied:
for example if we move public and private zone into module, then the module needs 2 variables... `base_domain` `cluster_domain`. And if I want to switch off the public zone records, I will have to pass empty `base_domain` IMO i don't like this indicator.

on the other hand if we pass zone ids, in the root module i pass empty zone id for public zone and that IMO is better indicator.

thoughts?

staebler replied:
If we were to make the public zone records optional, then there would need to be a variable coming into terraform to cover that configuration. That same variable would be passed down to the route53 module. I would not suggest overloading the semantics of `base_domain` to use it to indicate whether the public zone records are off or on. In my opinion, adding an additional variable to the route53 module that explicitly configures whether the public zone records are enabled is a worthwhile trade-off for keeping route53-specific resources contained within the route53 module.

abhinavdahiya replied:
done with https://github.com/openshift/installer/pull/1280/commits/d4ce44e391bdcef7c476c887f6234ee741b888d7

abhinavdahiya commented:
> These changes make me happy.
> 
>     * Should the following use cluster_id?
>       
>         
>           
>             [installer/data/data/aws/bootstrap/main.tf](https://github.com/openshift/installer/blob/e9073fd3dcdef7595b7eca927b7944e7e1e76dcd/data/data/aws/bootstrap/main.tf#L37)
>           
>           
>                Line 37
>             in
>             [e9073fd](/openshift/installer/commit/e9073fd3dcdef7595b7eca927b7944e7e1e76dcd)
>           
>           
>           
>           
>       
>               
>                 
>                  name = "${var.cluster_name}-bootstrap-profile"

Fixed here https://github.com/openshift/installer/pull/1280/commits/d2d9bbc144a8eb9ae76a38c02ca7f7114ef0bf14

>     * `pkg/asset/machines/worker.go` is using the UUID instead of the human ID.

Fixed here https://github.com/openshift/installer/pull/1280/commits/1a2a37f8c23c315e3c107dc33e710426df6ef23c https://github.com/openshift/installer/pull/1280/commits/bda7b9450b40e9ee1aaedb272db75298ebc637c9 https://github.com/openshift/installer/pull/1280/commits/6e6d71fb7a2133466b24da2c42495af9ae4065ed

>     * I don't like that the term cluster ID is overloaded. It would be nice if one of the two uses were qualified in some way so that it is evident from the term which of the cluster IDs is being used.

I think it is fair to say the clusterID gived to 2 options a UUID that you can use where you absolutely want global uniquenes and HumanID where you care about uniqueness to only small level.

staebler commented:
> I think it is fair to say the clusterID gived to 2 options a UUID that you can use where you absolutely want global uniquenes and HumanID where you care about uniqueness to only small level.

Fair enough. If you think that it will always be completely obvious which specific kind of cluster ID is required for every parameter passed to every function, then I am satisfied.

staebler commented on data/data/config.tf:
```diff
@@ -65,6 +65,11 @@ EOF
+  description = <<EOF
+(internal) The OpenShift cluster id.
+
+This cluster id must be of max length 32 and must have only alphabumeric or hyphen characters.
```
The max length is 27, not 32.

abhinavdahiya replied:
done https://github.com/openshift/installer/pull/1280/commits/b156efaf37b835972ccd68b9ec07279f05cdad24

staebler commented on data/data/config.tf:
```diff
@@ -65,6 +65,11 @@ EOF
+  description = <<EOF
+(internal) The OpenShift cluster id.
+
+This cluster id must be of max length 32 and must have only alphabumeric or hyphen characters.
```
s/alphabumeric/alphanumeric/

abhinavdahiya replied:
done with 144ab05a5

staebler commented on pkg/asset/installconfig/clusterid.go:
```diff
@@ -1,30 +1,70 @@
 }
+
+// generateHumanID take base and returns a ID that
+// - is of max length maxBaseLen + randomLen + 1
```
`is of max length maxNameLen`

abhinavdahiya replied:
Oh Yeah :) will fix that.

abhinavdahiya replied:
done with https://github.com/openshift/installer/pull/1280/commits/9f5c0c00348230fdf601bb466b2793638e852a8b

staebler commented on pkg/asset/installconfig/clusterid_test.go:
```diff
@@ -0,0 +1,33 @@
+			got := generateHumanID(test.input)
+			t.Log("HumanID", got)
+			assert.Equal(t, test.expLen, len(got))
+			assert.Equal(t, regexp.MustCompile("[^A-Za-z0-9-]").ReplaceAllString(got, "-"), got)
```
I would rather see the expected replacements written out explicitly in the test cases. This just ends up running the same code that is used in `generateHumanID`. If there is a bug in the replacement, this won't capture the bug.

abhinavdahiya replied:
I tried to make it more meaningful with https://github.com/openshift/installer/pull/1280/commits/9f5c0c00348230fdf601bb466b2793638e852a8b -> https://github.com/openshift/installer/pull/1280/commits/4650c151c932b38341c39cdf71658eed1e695679 [compare](https://github.com/openshift/installer/compare/79a821c2088b46314d44913e7f51a8fbfe58e99f..653a9b2e8e9080b6729fae26eb2bb45dc7e232ce)

abhinavdahiya commented:
looks like network operator error [e2e-aws](https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_installer/1280/pull-ci-openshift-installer-master-e2e-aws/4010):
```console
level=info msg="Waiting up to 30m0s for the cluster to initialize..."
level=fatal msg="failed to initialize the cluster: timed out waiting for the condition"
```
```json
        {
            "apiVersion": "config.openshift.io/v1",
            "kind": "ClusterOperator",
            "metadata": {
                "creationTimestamp": "2019-02-20T23:36:45Z",
                "generation": 1,
                "name": "network",
                "resourceVersion": "29767",
                "selfLink": "/apis/config.openshift.io/v1/clusteroperators/network",
                "uid": "628e2788-3568-11e9-b802-0efd314e1e72"
            },
            "spec": {},
            "status": {
                "conditions": [
                    {
                        "lastTransitionTime": "2019-02-20T23:36:53Z",
                        "status": "False",
                        "type": "Failing"
                    },
                    {
                        "lastTransitionTime": "2019-02-20T23:36:53Z",
                        "message": "DaemonSet \"openshift-multus/multus\" is not available (awaiting 1 nodes)\nDaemonSet \"openshift-sdn/ovs\" is not available (awaiting 1 nodes)\nDaemonSet \"openshift-sdn/sdn\" is not available (awaiting 1 nodes)",
                        "reason": "Deploying",
                        "status": "True",
                        "type": "Progressing"
                    },
                    {
                        "lastTransitionTime": "2019-02-20T23:36:53Z",
                        "message": "DaemonSet \"openshift-multus/multus\" is not available (awaiting 1 nodes)\nDaemonSet \"openshift-sdn/ovs\" is not available (awaiting 1 nodes)\nDaemonSet \"openshift-sdn/sdn\" is not available (awaiting 1 nodes)",
                        "reason": "Deploying",
                        "status": "False",
                        "type": "Available"
                    }
                ],
                "extension": null,
                "version": ""
            }
        },
```
/retest

abhinavdahiya commented:
/retest

abhinavdahiya commented:
The CI is green :)

dgoodwin commented on pkg/asset/installconfig/clusterid.go:
```diff
@@ -1,30 +1,70 @@
+	UUID string
+
+	// HumanID is an identifier for the cluster that is more human friendly.
+	// This does not have
```
Partial comment missing here.

dgoodwin replied:
But if you have green CI then I wouldn't hold it up. :)

dgoodwin commented:
If we vendor this in after it merges, picking up new deprovision code, what will happen in an environment where there are pre-existing clusters, specifically with regard to "All resources are tagged with kubernetes.io/cluster/<cluster_id>: owned". Pre-existing clusters will still have name:owned, will they still deprovision cleanly?

dgoodwin commented on pkg/asset/cluster/aws/aws.go:
```diff
@@ -12,13 +12,8 @@ import (
-		},
+		Identifier: []map[string]string{{
+			fmt.Sprintf("kubernetes.io/cluster/%s", clusterID): "owned",
+		}},
```
This part right here looks like it will break deprovision for any pre-existing clusters. For better or worse, there are some. Is there any way we can work around that?

dgoodwin replied:
CC @jhernand

staebler commented:
> If we vendor this in after it merges, picking up new deprovision code, what will happen in an environment where there are pre-existing clusters, specifically with regard to "All resources are tagged with kubernetes.io/cluster/<cluster_id>: owned". Pre-existing clusters will still have name:owned, will they still deprovision cleanly?

@dgoodwin Does hive retain the metadata.json for clusters? Or does hive re-create the metadata.json when it comes time to tear down the cluster? If the metadata.json is retained, then the destroyer will continue to use the old tags, which will match for the pre-existing cluster.

dgoodwin commented:
> > If we vendor this in after it merges, picking up new deprovision code, what will happen in an environment where there are pre-existing clusters, specifically with regard to "All resources are tagged with kubernetes.io/cluster/<cluster_id>: owned". Pre-existing clusters will still have name:owned, will they still deprovision cleanly?
> 
> @dgoodwin Does hive retain the metadata.json for clusters? Or does hive re-create the metadata.json when it comes time to tear down the cluster? If the metadata.json is retained, then the destroyer will continue to use the old tags, which will match for the pre-existing cluster.

We have the metadata.json but just for safe keeping, unfortunately we vendor deprovision code and use it in our hiveutil deprovision which doesn't use metadata, just a list of tags. With that in mind though I realize now we can probably solve it there, just add both formats of the tag to the tags to scrub, and we can remove one once we're sure everything is cleaned out of opshive.

crawford commented:
/test e2e-aws

abhinavdahiya commented:
> See [8a37f72](https://github.com/openshift/installer/commit/8a37f72b5877b0623348334c7a9e40f94a2cab10) and [f56c341](https://github.com/openshift/installer/commit/f56c34153d9976cd40e9dc68ca6ad61683b04007) about not putting `kubernetes.io/cluster/...` on bootstrap resources except the instance. Do you feel like the accidental-adoption concern is ungrounded?

the kubelet [service](https://github.com/openshift/installer/blob/c957fa46636b78189b5dbac076b7371f53fe8831/data/data/bootstrap/systemd/units/kubelet.service.template#L11-L22) on the bootstrap node has not cloud provider flags, therefore the error mentioned in [8a37f72](https://github.com/openshift/installer/commit/8a37f72b5877b0623348334c7a9e40f94a2cab10) shouldn't happen on bootstrap kubelet. @wking 

> I would like some more weigh-in on whether tagging the bootstrap resources with `kubernetes.io/cluster/...` has potential side effects.

using the `kubenetes.io/cluster/...` shouldn't have side effects as AWS controllers use it to find its internal_cluster_id and not more.

staebler commented on pkg/types/clustermetadata.go:
```diff
@@ -9,8 +9,12 @@ import (
+	// clusterID is a globally unique ID that is used to identify an Openshift cluster.
+	ClusterID string `json:"clusterID"`
+	// infraID is an ID that is used to identify cloud resources created by the installer.
+	InfraID                 string `json:"infraID"`
```
Now that we are using a different term for the human-readable cluster ID somewhere, is it worth revisiting whether it would be better to use that term everywhere the human-readable cluster ID is used?

abhinavdahiya replied:
`metadata.json` is an output to the users like hive, they don't care what installer used as ID, just that its the ID used to tag/name resources. 
`Human readable ID` IMO is an installer's internal impl and doesn't need to bleed into metadata.json

abhinavdahiya replied:
I updated the `ClusterID.HumanID` to `ClusterID.InfraID`

/cc @crawford @staebler

abhinavdahiya commented:
/test e2e-aws

abhinavdahiya commented:
/test e2e-aws

abhinavdahiya commented:
rebased around #1163

abhinavdahiya commented:
missed dropping variable during rebase :P

abhinavdahiya commented:
```
Failing tests:

[Conformance][Area:Networking][Feature:Router] The HAProxy router should expose prometheus metrics for a route [Suite:openshift/conformance/parallel/minimal]
```

router flaking
/retest

dgoodwin commented on pkg/asset/machines/libvirt/machinesets.go:
```diff
@@ -32,8 +31,8 @@ func MachineSets(clusterID string, config *types.InstallConfig, pool *types.Mach
-	provider := provider(clustername, config.Networking.MachineCIDR.String(), platform, userDataSecret)
-	name := fmt.Sprintf("%s-%s-%d", clustername, pool.Name, 0)
+	provider := provider(clusterID, config.Networking.MachineCIDR.String(), platform, userDataSecret)
+	name := fmt.Sprintf("%s-%s-%d", clusterID, pool.Name, 0)
```
What should Hive be passing when we call this function? This looks like it's still cluster ID, but should probably be infraID?

dgoodwin replied:
N/M found where it was passed in, we're sending infraID, var could probably use a rename though.
